presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkDiskPageInput.java

```diff
@@ -0,0 +1,148 @@
+            TempStorage tempStorage,
+            TempDataOperationContext tempDataOperationContext,
+            PrestoSparkBroadcastTableCacheManager prestoSparkBroadcastTableCacheManager,
+            StageId stageId)
+    {
+        this.pagesSerde = requireNonNull(pagesSerde, "pagesSerde is null");
+        this.serializedPageIterators = broadcastTableFilesInfo.stream()
+                .map(tableFiles -> loadBroadcastTable(tableFiles, tempStorage, tempDataOperationContext, prestoSparkBroadcastTableCacheManager, stageId))
+                .collect(Collectors.toList());
+    }
+
+    private synchronized Iterator<PrestoSparkSerializedPage> loadBroadcastTable(
+            List<PrestoSparkDiskPage> broadcastTaskFilesInfo,
+            TempStorage tempStorage,
+            TempDataOperationContext tempDataOperationContext,
+            PrestoSparkBroadcastTableCacheManager prestoSparkBroadcastTableCacheManager,
+            StageId stageId)
+    {
+        // Empty broadcast table
+        if (broadcastTaskFilesInfo.size() == 0) {
+            return Collections.emptyIterator();
+        }
+
+        List<PrestoSparkSerializedPage> pages = new ArrayList<>();
+
+        // Try to get table from cache
+        if (prestoSparkBroadcastTableCacheManager.isBroadcastTableCachedForStage(stageId)) {
+            pages = prestoSparkBroadcastTableCacheManager.getBroadcastTableForStage(stageId);
+        }
+        else {
+            try {
+                shuffle(broadcastTaskFilesInfo);
+                for (PrestoSparkTaskOutput taskFileInfo : broadcastTaskFilesInfo) {
+                    PrestoSparkDiskPage prestoSparkDiskPage = (PrestoSparkDiskPage) taskFileInfo;
+                    TempStorageHandle tempStorageHandle = tempStorage.deserialize(prestoSparkDiskPage.getSerializedStorageHandle());
+                    log.info("Reading path: " + tempStorageHandle.toString());
+                    InputStream inputStream = tempStorage.open(tempDataOperationContext, tempStorageHandle);
+                    Iterator<SerializedPage> pagesIterator = readSerializedPages(new InputStreamSliceInput(inputStream));
+                    int pageIndex = 0;
+                    while (pagesIterator.hasNext()) {
+                        SerializedPage page = pagesIterator.next();
+                        long checksum = computeChecksum(page.getSlice().byteArray());
+                        if (prestoSparkDiskPage.getPageChecksum().get(pageIndex) != checksum) {
+                            throw new RuntimeException("Disk page checksum does not match. " +
+                                    "Data seems to be corrupted on disk for file " + tempStorageHandle.toString());
+                        }
+                        pages.add(toPrestoSparkSerializedPage(page));
+                        pageIndex++;
+                    }
+
+                    // Put the fetched pages in cache
+                    prestoSparkBroadcastTableCacheManager.cache(stageId, pages);
```

Comment:
Since now we have control over caching we need to reserve the memory used by the cache using the Presto memory reservation mechanism. It might be somehow tricky though. Presto offers `LocalMemoryContext` that is "per operator". There will be several instances of the `PrestoSparkRemoteSourceOperator` for a single plan node, though the memory for caching should be accounted only once.